From d7b9ab991be9d91bc2e3f4bde97e3df04f04684f Mon Sep 17 00:00:00 2001
From: Hemant Kumar <hemantk@codeaurora.org>
Date: Thu, 29 May 2014 16:05:51 -0700
Subject: [PATCH] usb: core: Fix use after free for hub usb device

There is a possibility of root hub usb device being
freed as part of controller driver unbind and hub
thread trying to access root hub usb device. Hence
increment the usb device reference count before
accessing it by hub thread.

Change-Id: Icf1297452e55a4ca0acd67075320f7a32929a1b3
CRs-Fixed: 671581
Signed-off-by: Hemant Kumar <hemantk@codeaurora.org>
---
 drivers/usb/core/hub.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/usb/core/hub.c b/drivers/usb/core/hub.c
index 76b9f3b..b7ccd05 100644
--- a/drivers/usb/core/hub.c
+++ b/drivers/usb/core/hub.c
@@ -4764,6 +4764,12 @@ static void hub_events(void)
 				(u16) hub->change_bits[0],
 				(u16) hub->event_bits[0]);
 
+		/* make sure hdev is not freed before accessing it */
+		if (hub->disconnected)
+			goto hub_disconnected;
+		else
+			usb_get_dev(hdev);
+
 		/* Lock the device, then check to see if we were
 		 * disconnected while waiting for the lock to succeed. */
 		usb_lock_device(hdev);
@@ -4968,6 +4974,8 @@ static void hub_events(void)
 		usb_autopm_put_interface(intf);
  loop_disconnected:
 		usb_unlock_device(hdev);
+		usb_put_dev(hdev);
+ hub_disconnected:
 		kref_put(&hub->kref, hub_release);
 
         } /* end while (1) */
-- 
1.8.2.1

